# ICD341 Laboratory 4 — Profiling and Benchmarking Python Code

## Topic
Profiling, benchmarking, computational performance, and bottleneck identification.

## Objectives

The objectives of this laboratory are to:

- Explain benchmarking.
- Use `%timeit` to measure execution time.
- Compare alternative Python implementations.
- Identify computational bottlenecks.
- Install and use `line_profiler`.
- Explain why performance matters in statistical computing.

## Overview

This laboratory investigates the performance of Python and NumPy implementations using benchmarking and profiling techniques.

The laboratory compares loop-based and vectorized approaches and examines how execution time changes as dataset size increases.

In [1]:
import numpy as np

## Part A — `%timeit`

`%timeit` is a Jupyter/IPython magic command used to measure the execution time of Python code.

It runs the expression multiple times and reports statistics about the execution time. This provides a more reliable performance measurement than timing a single execution.

The following example benchmarks a NumPy vectorized operation on one million values.

In [2]:
x = np.arange(1_000_000)

%timeit x ** 2

2.85 ms ± 138 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


## Part B — Comparing a Python Loop with NumPy Vectorization

This section compares two implementations of the same mathematical operation.

The first implementation uses a traditional Python `for` loop, while the second uses NumPy vectorization.

Both functions calculate the square of every element in an array. `%timeit` is used to measure their execution times on the current computer.

In [3]:
def square_loop(x):
    result = []
    for value in x:
        result.append(value ** 2)
    return result

In [4]:
%timeit square_loop(x)

145 ms ± 8.89 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


In [5]:
def square_numpy(x):
    return x ** 2

In [6]:
%timeit square_numpy(x)

3.57 ms ± 546 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


### Task 1 — Benchmark Comparison

| Method | Measured Time |
|---|---:|
| Python loop | 145 ms ± 8.89 ms |
| NumPy vectorization | 3.57 ms ± 546 μs |

#### Questions

**1. Which implementation was faster?**

The NumPy vectorized implementation was faster on this computer. Its measured mean execution time was 3.57 ms compared with 145 ms for the Python loop.

**2. How many times faster was it?**

Using the mean benchmark times:

Speed-up = Loop time / NumPy time

Speed-up = 145 / 3.57 ≈ **40.6×**

Therefore, the NumPy implementation was approximately 40.6 times faster in this benchmark.

**3. Why might NumPy be faster?**

NumPy performs array operations using optimized numerical routines implemented largely in compiled code. This reduces the overhead associated with executing a Python `for` loop for each individual element.

The exact speed-up depends on the computer, Python version, NumPy version, array size, and system conditions. Therefore, benchmark results should be measured rather than assumed.

## Part C — Benchmarking Different Array Sizes

The amount of data can affect execution time. This section benchmarks the same vectorized operation using arrays containing 100, 10,000, and 1,000,000 elements.

The purpose is to observe how execution time changes as the dataset becomes larger.

In [7]:
x1 = np.arange(100)
x2 = np.arange(10_000)
x3 = np.arange(1_000_000)

In [8]:
%timeit x1 ** 2

649 ns ± 21.9 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


In [9]:
%timeit x2 ** 2

8.14 μs ± 822 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [10]:
%timeit x3 ** 2

3.2 ms ± 273 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


### Task 2 — Effect of Dataset Size

The measured execution times were:

| Array Size | Measured Time |
|---:|---:|
| 100 | 649 ns ± 21.9 ns |
| 10,000 | 8.14 μs ± 822 ns |
| 1,000,000 | 3.20 ms ± 273 μs |

As the dataset becomes larger, the running time generally increases because NumPy must perform the operation on more elements.

The results demonstrate that dataset size is an important factor when considering computational performance. Although vectorized NumPy operations are efficient, processing very large datasets still requires more computational work than processing small datasets.

The exact execution time also depends on factors such as the computer hardware, software environment, memory, and system activity.

## Part D — Benchmarking Statistical Calculations

Statistical computing often involves operations such as calculating means, standard deviations, and sums over large datasets.

This section measures the execution time of several NumPy statistical operations using one million normally distributed observations.

In [11]:
data = np.random.normal(size=1_000_000)

In [12]:
%timeit np.mean(data)

1.38 ms ± 132 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [13]:
%timeit np.std(data)

6.09 ms ± 197 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [14]:
%timeit np.sum(data)

990 μs ± 195 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


### Task 3 — Statistical Calculation Performance

The measured execution times were:

| Operation | Measured Time |
|---|---:|
| `np.mean(data)` | 1.38 ms ± 132 μs |
| `np.std(data)` | 6.09 ms ± 197 μs |
| `np.sum(data)` | 990 μs ± 195 μs |

#### 1. Which operation was fastest on my computer?

`np.sum(data)` was the fastest operation, with a measured mean execution time of **990 μs (0.990 ms)**.

`np.mean(data)` took **1.38 ms**, while `np.std(data)` took **6.09 ms**.

#### 2. Why should performance be measured rather than assumed?

Performance should be measured because different operations perform different amounts of computational work, and their execution times can also depend on the computer hardware, software environment, dataset size, memory, and system activity.

Benchmarking provides actual measurements instead of relying on assumptions about which operation will be faster.

In statistical computing, measuring performance is particularly important when working with large datasets or computationally expensive calculations.

In [2]:
%load_ext line_profiler

In [3]:
from profiling_example import calculate_statistics

In [4]:
import numpy as np

data = np.random.normal(size=1_000_000)

In [5]:
from profiling_example import calculate_statistics

In [5]:
calculate_statistics(data)

(np.float64(-923.6409292383664),
 np.float64(-0.0009236409292383665),
 np.float64(0.9989982376830152))

In [8]:
%lprun -f calculate_statistics calculate_statistics(data)

Timer unit: 1e-07 s

Total time: 0.0110089 s
File: C:\Users\asus\ICD341\labs\profiling_example.py
Function: calculate_statistics at line 3

Line #      Hits         Time  Per Hit   % Time  Line Contents
     3                                           def calculate_statistics(x):
     4         1      12777.0  12777.0     11.6      total = np.sum(x)
     5         1      14361.0  14361.0     13.0      mean = np.mean(x)
     6         1      82749.0  82749.0     75.2      variance = np.var(x)
     7         1        178.0    178.0      0.2      result = np.sqrt(variance)
     8         1         24.0     24.0      0.0      return total, mean, result

### Task 3 — Line Profiling Results

The `line_profiler` results show that the function took approximately **0.0110089 seconds (11.01 ms)** in total.

| Operation | Time | Percentage of Total |
|---|---:|---:|
| `np.sum(x)` | 1.2777 ms | 11.6% |
| `np.mean(x)` | 1.4361 ms | 13.0% |
| `np.var(x)` | 8.2749 ms | 75.2% |
| `np.sqrt(variance)` | 0.0178 ms | 0.2% |
| `return` | 0.0024 ms | 0.0% |

#### 1. Which line takes the most time?

The line:

`variance = np.var(x)`

takes the most time. It accounts for **75.2%** of the total measured execution time.

#### 2. How much time does each line take?

The profiler measured approximately:

- `np.sum(x)`: **1.2777 ms**
- `np.mean(x)`: **1.4361 ms**
- `np.var(x)`: **8.2749 ms**
- `np.sqrt(variance)`: **0.0178 ms**
- `return`: **0.0024 ms**

#### 3. Which part might need optimization?

The variance calculation is the main computational bottleneck in this function because it accounts for the largest proportion of the execution time.

Profiling helps identify this bottleneck so that optimization efforts can focus on the part of the program that consumes the most computational time rather than optimizing lines that have very little impact.

The exact profiling times are specific to this computer and execution, so they should be treated as measured results rather than universal values.

## Part F — Identifying an Inefficient Function

This section investigates a deliberately inefficient function that calculates the sum of squared values using a Python loop.

The function is profiled to identify which operations consume the most execution time. It is then compared with a vectorized NumPy implementation.

In [9]:
def slow_function(x):
    result = []
    for value in x:
        result.append(value ** 2)
    total = sum(result)
    return total

In [7]:
x = np.arange(1_000_000)

In [10]:
%lprun -f slow_function slow_function(x)

Timer unit: 1e-07 s

Total time: 0.720012 s
File: C:\Users\asus\AppData\Local\Temp\ipykernel_21648\3756814506.py
Function: slow_function at line 1

Line #      Hits         Time  Per Hit   % Time  Line Contents
     1                                           def slow_function(x):
     2         1         29.0     29.0      0.0      result = []
     3   1000001    3300938.0      3.3     45.8      for value in x:
     4   1000000    3606491.0      3.6     50.1          result.append(value ** 2)
     5         1     292636.0 292636.0      4.1      total = sum(result)
     6         1         24.0     24.0      0.0      return total

In [11]:
def fast_function(x):
    return np.sum(x ** 2)

In [12]:
%timeit slow_function(x)
%timeit fast_function(x)

222 ms ± 47.1 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
4.56 ms ± 488 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)


## Task 4 — Comparing the Inefficient and Optimized Functions

The deliberately inefficient `slow_function()` uses a Python `for` loop to square each element individually, stores the results in a Python list, and then calculates the total using `sum()`.

The optimized `fast_function()` uses NumPy vectorization and `np.sum()` to perform the same calculation.

### Benchmark Results

| Function | Measured Time |
|---|---:|
| `slow_function(x)` | 222 ms ± 47.1 ms |
| `fast_function(x)` | 4.56 ms ± 488 μs |

### Speed-up

Speed-up = Slow function time / Fast function time

Speed-up = 222 / 4.56 ≈ **48.7×**

Therefore, the optimized NumPy implementation was approximately **48.7 times faster** than the Python-loop implementation in this benchmark.

### Why is `fast_function()` more efficient?

The `slow_function()` performs element-by-element operations in a Python `for` loop and repeatedly appends values to a Python list. These operations introduce significant Python-level overhead.

The `fast_function()` uses NumPy vectorization:

```python
np.sum(x ** 2)

## Part G — Mini Statistical Computing Investigation

This section compares two approaches for calculating the sum of squared deviations from the mean.

The first approach uses a Python `for` loop, while the second uses NumPy vectorization.

The experiment uses a large dataset containing 5,000,000 normally distributed observations. The two implementations will be benchmarked to determine the performance difference between loop-based and vectorized statistical computation.

In [13]:
np.random.seed(42)

data = np.random.normal(
    loc=100,
    scale=20,
    size=5_000_000
)

In [14]:
def squared_deviations_loop(x):
    mean = np.mean(x)
    total = 0
    for value in x:
        total += (value - mean) ** 2
    return total

In [15]:
def squared_deviations_numpy(x):
    mean = np.mean(x)
    return np.sum((x - mean) ** 2)

In [16]:
%timeit -n 1 -r 3 squared_deviations_loop(data)

1.21 s ± 82.3 ms per loop (mean ± std. dev. of 3 runs, 1 loop each)


In [17]:
%timeit -n 1 -r 3 squared_deviations_numpy(data)

44.2 ms ± 2.52 ms per loop (mean ± std. dev. of 3 runs, 1 loop each)


### Part G — Results and Analysis

The two implementations were benchmarked using the same dataset containing 5,000,000 normally distributed observations.

| Approach | Measured Time |
|---|---:|
| Python loop | 1.21 s ± 82.3 ms |
| NumPy vectorization | 44.2 ms ± 2.52 ms |

#### 1. Which approach was faster?

The NumPy vectorized approach had the lower measured execution time in this benchmark.

#### 2. What was the measured time for each approach?

- Python loop: **1.21 seconds ± 82.3 ms**
- NumPy vectorization: **44.2 ms ± 2.52 ms**

#### 3. Speed-up

Speed-up = Loop time / NumPy time

Speed-up = 1.21 / 0.0442 ≈ **27.4×**

Therefore, the NumPy implementation was approximately **27.4 times faster** in this benchmark.

#### 4. Why can vectorization produce a large performance difference?

The loop-based implementation performs the calculation one element at a time using Python-level iteration. This introduces overhead for processing each individual element.

The NumPy implementation performs the array operation using optimized numerical routines operating on the entire array. This reduces Python-level overhead and makes better use of optimized compiled code.

The large difference becomes particularly noticeable when processing millions of observations.

#### 5. How does profiling help decide where to optimize a statistical program?

Profiling identifies which parts of a program consume the most execution time. This allows optimization efforts to focus on actual computational bottlenecks rather than making unnecessary changes to code that has little effect on overall performance.

In this laboratory, profiling showed that the loop and list operations were major contributors to the execution time of `slow_function()`. Benchmarking then demonstrated the performance improvement obtained by replacing the loop-based approach with NumPy vectorization.

### Conclusion

This investigation demonstrates that performance measurement is important in statistical computing. Python loops can introduce significant overhead when processing large datasets, while NumPy vectorization can substantially reduce execution time.

The measured results are specific to this computer and software environment, so performance should be benchmarked rather than assumed.